### PR TITLE
[SharedOffloadRegion] Align blocks to page-size  

### PR DESCRIPTION
**Pull request by varun-sundar-rabindranath**
## Purpose
Align blocks in SharedOffloadRegion to page_size so that O_DIRECT succeeds.

Changes: 
 - Update CPUOffloadingSpec to account for alignment
 - Update SharedOffloadRegion to compute aligned row_strides
 - Update test_fs_tier.py to use SharedOffloadRegion for robust testing and testing the interplay between fs_tier and SharedOffloadRegion 

Interface change:
  - CPUOffloadingSpec constructor inputs a block_size_alignment integer that defaults to 1
  - SharedOffloadRegion directly consumes the user specified cpu_bytes_to_use.

## Test Plan
 Run `pytest -s tests/v1/kv_offload/test_fs_tier.py` multiple times locally.

## Test Result
Test passes.

### CODE REVIEW

**Comment by orozery** [comment]
Thanks @varun-sundar-rabindranath !
This fixes the test, but leaves the possible underlying issue.
Once we fix the underlying issue, this test would actually prove we fixed it.

**Comment by varun-sundar-rabindranath** [comment]
> Thanks @varun-sundar-rabindranath ! This fixes the test, but leaves the possible underlying issue. Once we fix the underlying issue, this test would actually prove we fixed it.

Thanks for taking a look @orozery. The fixes are the root cause of the test failure. The current set of tests dont invoke the SharedOffloadRegion. These are targeted tests that test the fs_tier directly. 
 
IMO Making sure that the CPU backing tensor is always aligned should be a separate test. what do you think ?

**Comment by varun-sundar-rabindranath** [comment]
> > Thanks @varun-sundar-rabindranath ! This fixes the test, but leaves the possible underlying issue. Once we fix the underlying issue, this test would actually prove we fixed it.
> 
> Thanks for taking a look @orozery. The fixes are the root cause of the test failure. The current set of tests dont invoke the SharedOffloadRegion. These are targeted tests that test the fs_tier directly.
> 
> IMO Making sure that the CPU backing tensor is always aligned should be a separate test. what do you think ?

Hi @orozery  I have updated the tests to use SharedOffloadRegion directly for better tests and have updated SharedOffloadRegion to always align its rows to page size boundaries. PTAL! Thanks.

File: vllm/v1/kv_offload/cpu/shared_offload_region.py [hunk 1]
```diff
@@ -44,15 +69,23 @@ def __init__(
         num_workers: int,
         cpu_page_size: int,
     ) -> None:
+        assert total_size_bytes >= num_blocks * cpu_page_size * num_workers, (
+            f"{total_size_bytes=} < {num_blocks=} * {cpu_page_size=} * {num_workers=}"
+        )
+
         self.page_size = mmap.PAGESIZE
+        self.num_blocks = num_blocks
+        self.total_size_bytes, self._row_stride = self._maybe_update_buffer_size(
```
Comment (orozery) [inline vllm/v1/kv_offload/cpu/shared_offload_region.py]: This potentially violates the user's `cpu_bytes_to_use`, allocating more than the user allowed.
I think we want to add an alignment classvar in `cpu/spec.py`, which will be overrided in `tiering/spec.py`:
```
  # CPUOffloadingSpec
  class CPUOffloadingSpec(OffloadingSpec):
      CPU_PAGE_SIZE_ALIGNMENT = 1

      def __init__(self, vllm_config, kv_cache_config):
          ...
          kv_bytes_per_offloaded_block = kv_bytes_per_block * self.block_size_factor
          self.cpu_page_size_per_worker = round_up(
              kv_bytes_per_offloaded_block // world_size,
              self.CPU_PAGE_SIZE_ALIGNMENT,
          )
          self.num_blocks = (
              int(cpu_bytes_to_use) // (self.cpu_page_size_per_worker * world_size)
              if self.cpu_page_size_per_worker > 0
              else 0
          )
          ...

  # TieringOffloadingSpec
  class TieringOffloadingSpec(CPUOffloadingSpec):
      CPU_PAGE_SIZE_ALIGNMENT = SharedOffloadRegion.PAGE_SIZE_ALIGNMENT
      ...
```

File: tests/v1/kv_offload/test_fs_tier.py [hunk 2]
```diff
@@ -90,15 +93,49 @@ def drain(tier: FileSystemTierManager, max_rounds: int = 40) -> list:
...
 
 @pytest.fixture
 def fs_tier(tmp_path):
-    tensor = torch.zeros((4, _BLOCK_ELEMENTS), dtype=_DTYPE)
-    mock_view = memoryview(tensor.numpy())
+    region, tensor, mock_view = _make_region_tensor_and_view(
+        num_blocks=4,
+        block_elements=_BLOCK_ELEMENTS,
+        instance_prefix="test-fs-tier",
+    )
```
Comment (orozery) [inline tests/v1/kv_offload/test_fs_tier.py]: Assuming we move the alignment code further up to `spec.py`, let's go back to a simple tensor allocation here, but using an aligned page size.

**Review by orozery — CHANGES_REQUESTED** [review]
Thanks @varun-sundar-rabindranath !
Can you please change the PR title and description to reflect it aligns CPU pages?

File: vllm/v1/kv_offload/cpu/shared_offload_region.py (hunk 1 above)
Comment (varun-sundar-rabindranath) [reply vllm/v1/kv_offload/cpu/shared_offload_region.py]: Nice catch. I agree that `cpu_bytes_to_use` should be respected. 

Concern: 
Expanding individual `cpu_page_size_per_worker` looks like it'll break a invariants and introduce some hard-to-catch bugs. i.e. We are going from, 
```
  B0 |<--- B0 W0 ---><--- B0 W1 ---><--- B0 W2 --->| 
  B1 |<--- B1 W0 ---><--- B1 W1 ---><--- B1 W2 --->| 
  B2 |<--- B2 W0 ---><--- B2 W1 ---><--- B2 W2 --->| 
  ...
 
where Bi - Block i ; Wj - Worker j
```
to 
```
  B0 |<--- B0 W0 ---***pad***><--- B0 W1 ---***pad***><--- B0 W2 ---***pad***>| 
  B1 |<--- B1 W0 ---***pad***><--- B1 W1 ---***pad***><--- B1 W2 ---***pad***>| 
  B2 |<--- B2 W0 ---***pad***><--- B2 W1 ---***pad****><--- B2 W2 ---***pad***>| 
  ...
 
where Bi - Block i ; Wj - Worker j
```
One example is the assert in `cpu <-> gpu` transfer 
             `assert cpu_page_size == gpu_page_size * block_size_factor`
  https://github.com/vllm-project/vllm/blob/52a31ccecca203effd490a5b496dc5f8d9496654/vllm/v1/kv_offload/cpu/gpu_worker.py#L154
This padding will have to be plumbed through and handled correctly. 

Instead I propose doing, 
```
  B0 |<--- B0 W0 ---><--- B0 W1 ---><--- B0 W2 --->***pad***| 
  B1 |<--- B1 W0 ---><--- B1 W1 ---><--- B1 W2 --->***pad***| 
  B2 |<--- B2 W0 ---><--- B2 W1 ---><--- B2 W2 --->***pad***| 
  ...
```
which is a looser constraint and can be handled directly in SharedOffloadRegion. Respecting cpu_bytes_to_use can be handled by, 
 - Allocating less num_blocks in `CPUOffloadingSpec` when padding is involved (communicated via CPU_PAGE_SIZE_ALIGNMENT classvar or a constructor arg)
 - passing `cpu_bytes_to_use` directly to SharedOffloadRegion.
 - And introducing padding in SharedOffloadRegion that respects alignment and `cpu_bytes_to_use`

File: tests/v1/kv_offload/test_fs_tier.py (hunk 2 above)
Comment (varun-sundar-rabindranath) [reply tests/v1/kv_offload/test_fs_tier.py]: Keeping this as the new set of changes still updates SharedOffloadRegion directly. PTAL. Thanks 🙌

File: vllm/v1/kv_offload/cpu/spec.py [hunk 3]
```diff
@@ -28,22 +34,31 @@ def __init__(self, vllm_config: VllmConfig, kv_cache_config: KVCacheConfig):
...
+            kv_bytes_per_offloaded_block = kv_bytes_per_block * self.block_size_factor
+            aligned_kv_bytes_per_offloaded_block = round_up(
+                kv_bytes_per_offloaded_block, self.block_size_alignment
+            )
+            assert self.cpu_bytes_to_use >= aligned_kv_bytes_per_offloaded_block, (
+                f"CPU space insufficient for offloading. {self.cpu_bytes_to_use=} "
+                f"{kv_bytes_per_offloaded_block=} "
+                f"{aligned_kv_bytes_per_offloaded_block=} "
+                f"{self.block_size_alignment=}"
+            )
```
Comment (orozery) [inline vllm/v1/kv_offload/cpu/spec.py]: Why do we need this assert?

File: vllm/v1/kv_offload/cpu/spec.py [hunk 4]
```diff
@@ -19,7 +20,12 @@
 
 
 class CPUOffloadingSpec(OffloadingSpec):
-    def __init__(self, vllm_config: VllmConfig, kv_cache_config: KVCacheConfig):
+    def __init__(
+        self,
+        vllm_config: VllmConfig,
+        kv_cache_config: KVCacheConfig,
+        block_size_alignment: int = 1,
```
Comment (orozery) [inline vllm/v1/kv_offload/cpu/spec.py]: Let's use a classvar instead of introducing a new __init__ param

File: vllm/v1/kv_offload/cpu/spec.py [hunk 5]
```diff
@@ -28,22 +34,31 @@ def __init__(self, vllm_config: VllmConfig, kv_cache_config: KVCacheConfig):
...
+        self.block_size_alignment = block_size_alignment
         assert kv_cache_config is not None
         if kv_cache_config.num_blocks > 0:
+            # calculate kv_bytes_per_offloaded_block and num_blocks
             total_gpu_kv_bytes = sum(t.size for t in kv_cache_config.kv_cache_tensors)
             kv_bytes_per_block = (
                 total_gpu_kv_bytes // kv_cache_config.num_blocks
             ) * vllm_config.parallel_config.world_size
+            kv_bytes_per_offloaded_block = kv_bytes_per_block * self.block_size_factor
+            aligned_kv_bytes_per_offloaded_block = round_up(
```
Comment (orozery) [inline vllm/v1/kv_offload/cpu/spec.py:52]: Let's save the aligned value to `self.kv_bytes_per_offloaded_block` and use this to pass in to `SharedOffloadRegion` (instead of `num_workers`).
This will be used as the stride.
We won't need all of the other changes to `SharedOffloadRegion`.

**Review by orozery — CHANGES_REQUESTED** [review]

File: vllm/v1/kv_offload/cpu/spec.py (hunk 3 above)
Comment (varun-sundar-rabindranath) [reply vllm/v1/kv_offload/cpu/spec.py]: This assert is making sure that the num_blocks is not 0. It indicates that the user should increase `cpu_bytes_to_use` to run with CPU offloading enabled.

File: vllm/v1/kv_offload/cpu/spec.py (hunk 4 above)
Comment (varun-sundar-rabindranath) [reply vllm/v1/kv_offload/cpu/spec.py]: Updated to introduce a classvar.
I am a bit uncomfortable with the semantic difference between `CPUOffloadingSpec.BLOCK_SIZE_ALIGNMENT` vs `self.BLOCK_SIZE_ALIGNMENT` seems like it increases the surface for bugs. what do you think ?

File: vllm/v1/kv_offload/cpu/spec.py (hunk 5 above)
Comment (varun-sundar-rabindranath) [reply vllm/v1/kv_offload/cpu/spec.py:52]: Done. Thanks for the suggestion.

File: vllm/v1/kv_offload/cpu/spec.py [hunk 6]
```diff
@@ -28,25 +35,36 @@ def __init__(self, vllm_config: VllmConfig, kv_cache_config: KVCacheConfig):
...
+            )
+            self.kv_bytes_per_offloaded_block_pad = round_up(
+                self.kv_bytes_per_offloaded_block, self.BLOCK_SIZE_ALIGNMENT
+            )
+            assert int(cpu_bytes_to_use) >= self.kv_bytes_per_offloaded_block_pad, (
+                f"CPU space insufficient for offloading. {cpu_bytes_to_use=} "
+                f"{self.kv_bytes_per_offloaded_block=} "
+                f"{self.kv_bytes_per_offloaded_block_pad=} "
+                f"{self.BLOCK_SIZE_ALIGNMENT=}"
+            )
```
Comment (orozery) [inline vllm/v1/kv_offload/cpu/spec.py]: Same as the previous review: Why do we need this assert?

File: vllm/v1/kv_offload/cpu/spec.py [hunk 7]
```diff
@@ -28,25 +35,36 @@ def __init__(self, vllm_config: VllmConfig, kv_cache_config: KVCacheConfig):
...
-        self.num_blocks = (
-            int(cpu_bytes_to_use) // kv_bytes_per_offloaded_block
-            if kv_bytes_per_offloaded_block > 0
-            else 0
-        )
+
+            self.kv_bytes_per_offloaded_block = (
+                kv_bytes_per_block * self.block_size_factor
+            )
+            self.kv_bytes_per_offloaded_block_pad = round_up(
```
Comment (orozery) [inline vllm/v1/kv_offload/cpu/spec.py]: Can we save just this variable to `self` and call it `kv_bytes_per_offloaded_block`? 
We can add a comment explaining this size possibly contains padding.

File: tests/v1/kv_offload/test_fs_tier.py [hunk 8]
```diff
@@ -90,15 +94,48 @@ def drain(tier: FileSystemTierManager, max_rounds: int = 40) -> list:
     return results
 
 
+def _make_region_tensor_and_view(
```
Comment (orozery) [inline tests/v1/kv_offload/test_fs_tier.py]: I prefer we keep the unit tests here independent of shared_offload_region.py.
Let's just fix the allocation to be page aligned.

**Review by orozery — CHANGES_REQUESTED** [review]

File: vllm/v1/kv_offload/cpu/spec.py (hunk 6 above)
Comment (varun-sundar-rabindranath) [reply vllm/v1/kv_offload/cpu/spec.py]: I replied in the previous review - https://github.com/vllm-project/vllm/pull/43689#discussion_r3318284972

File: vllm/v1/kv_offload/cpu/spec.py (hunk 7 above)
Comment (varun-sundar-rabindranath) [reply vllm/v1/kv_offload/cpu/spec.py]: having a single `self.kv_bytes_per_offloaded_block` was my first approach. 
Then I saw cpu_page_size_per_worker being computed from `kv_bytes_per_offloaded_block`. I then introduced `kv_bytes_per_offloaded_block_pad` to avoid confusion and for clarity.

File: tests/v1/kv_offload/test_fs_tier.py (hunk 8 above)
Comment (varun-sundar-rabindranath) [reply tests/v1/kv_offload/test_fs_tier.py]: Done.

File: vllm/v1/kv_offload/cpu/spec.py (hunk 7 above)
Comment (orozery) [reply vllm/v1/kv_offload/cpu/spec.py]: We don't need both, we need just the aligned one.
And I think the `_pad` suffix makes it too complicated.

File: vllm/v1/kv_offload/cpu/spec.py (hunk 6 above)
Comment (orozery) [reply vllm/v1/kv_offload/cpu/spec.py]: IMO This is a bit far fetched.
And if have a single block, what is it good for?

**Comment by orozery** [comment]
@varun-sundar-rabindranath can you please address the remaining nits?

File: vllm/v1/kv_offload/cpu/spec.py (hunk 6 above)
Comment (varun-sundar-rabindranath) [reply vllm/v1/kv_offload/cpu/spec.py]: I have reverted the assert. 
but, I believe it is better to be defensive in this case - for example I think having zero num_blocks will fail mmap in SharedOffloadRegion because num_blocks is zero. What do you think ? maybe we should handle it elsewhere.

File: vllm/v1/kv_offload/cpu/spec.py (hunk 7 above)
Comment (varun-sundar-rabindranath) [reply vllm/v1/kv_offload/cpu/spec.py]: reverted to have only `self.kv_bytes_per_offloaded_block`

**Comment by varun-sundar-rabindranath** [comment]
> @varun-sundar-rabindranath can you please address the remaining nits?

Hi @orozery  I have addressed the comments. PTAL! thanks 🙌

File: tests/v1/kv_offload/test_fs_tier.py [hunk 9]
```diff
@@ -90,14 +94,40 @@ def drain(tier: FileSystemTierManager, max_rounds: int = 40) -> list:
...
+    shift = page_size - alignment_offset
+    t = t[shift : shift + num_bytes]
+    return t.view(dtype).view(num_blocks, block_elements)
+
+
+def _page_aligned_rand_tensor(
+    num_blocks: int, block_elements: int, dtype: torch.dtype = _DTYPE
+) -> torch.Tensor:
+    rand_tensor = _page_aligned_zero_tensor(num_blocks, block_elements)
+    rand_tensor[:] = torch.rand(4, _BLOCK_ELEMENTS, dtype=dtype)
```
Comment (orozery) [inline tests/v1/kv_offload/test_fs_tier.py]:
```suggestion
    rand_tensor[:] = torch.rand(4, num_blocks, dtype=dtype)
```

File: tests/v1/kv_offload/test_fs_tier.py [hunk 10]
```diff
@@ -254,3 +284,51 @@ def test_store_load_data_integrity(fs_tier):
         assert torch.allclose(tensor[bid], expected[i]), (
             f"Block {bid} data mismatch after store+load"
         )
+
+
+@pytest.mark.parametrize("cpu_page_size", [mmap.PAGESIZE, mmap.PAGESIZE + 128])
```
Comment (orozery) [inline tests/v1/kv_offload/test_fs_tier.py]: Let's remove this test. It mixes `SharedOffloadRegion` which is out of scope.

File: vllm/v1/kv_offload/cpu/shared_offload_region.py [hunk 11]
```diff
@@ -35,24 +35,25 @@ class SharedOffloadRegion:
...
         self,
         instance_id: str,
-        total_size_bytes: int,
         num_blocks: int,
         rank: int | None,
-        num_workers: int,
+        kv_bytes_per_block: int,
         cpu_page_size: int,
     ) -> None:
         self.page_size = mmap.PAGESIZE
```
Comment (orozery) [inline vllm/v1/kv_offload/cpu/shared_offload_region.py:48]:
```suggestion
        self.page_size = mmap.PAGESIZE
        assert kv_bytes_per_block % self.page_size == 0
```

**Review by orozery — APPROVED** [review]
Thanks @varun-sundar-rabindranath !

**Comment by orozery** [comment]
@varun-sundar-rabindranath test failing:
https://buildkite.com/vllm/ci/builds/69346#019e8591-14d4-4f18-ac27-d3afcd3030d0
